### PR TITLE
[DeepSeek v3] Add grad mask and update MLA init

### PR DESCRIPTION
**Pull request by gagika**
# Description

When training DeepSeek-V3 671B with `load_balance_loss_weight > 0` deterministically NaNs on step 1 (per-layer MLA backward gradient overflows bf16). Adds two opt-in fixes that default to no-op:

1. **Per-token gradient mask** (`grad_mask_threshold`): zeros tokens whose feature-axis backward-RMS exceeds threshold.
2.  **Constant-std MLA init** (`mla_init_std`): N(0, std) with output proj scaled by 1/sqrt(2*num_decoder_layers). No effect when loading a checkpoint.

DeepSeek-V3 model config defaults these to 0.001 and 100.0. 

# Tests

- E2E test with random weights. https://cloudlogging.app.goo.gl/FU6FtqvXr6pegYv57
- E2E on mlperf ckpt0424: runs to lm_loss 3.622 over 84 steps. 
- Unit Tests in`tests/unit/grad_mask_utils_test.py`.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/grad_mask_utils.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+
+def _grad_mask_fwd(x: jax.Array, threshold: jax.Array):
+  return x, threshold
+
+
+def _grad_mask_bwd(threshold: jax.Array, g: jax.Array):
+  rms = jnp.sqrt(jnp.mean(jnp.square(g.astype(jnp.float32)), axis=-1, keepdims=True))
+  mask = (rms <= threshold).astype(g.dtype)
+  return (g * mask, jnp.zeros_like(threshold))
```
[the remark this answers is not in the document]
Comment (gagika) [reply src/maxtext/utils/grad_mask_utils.py]: done.

File: src/maxtext/configs/models/deepseek3-671b.yml [hunk 2]
```diff
@@ -44,6 +44,12 @@ qk_nope_head_dim: 128
 qk_rope_head_dim: 64
 v_head_dim: 128
 mscale: 1.0
+# Initialize MLA projections with N(0, std); output proj further scaled
+# by 1/sqrt(2*num_decoder_layers). No effect when loading a checkpoint.
+mla_init_std: 0.001
```
Comment (RissyRan) [inline src/maxtext/configs/models/deepseek3-671b.yml:49]: Let's move those out of model config but put in the base.ymal? Or we could add a comment around line #19?

File: src/maxtext/configs/base.yml [hunk 3]
```diff
@@ -396,6 +396,10 @@ qk_nope_head_dim: 128
 qk_rope_head_dim: 64
 v_head_dim: 128
 
+# Constant-std init for MLA proj; output proj scaled by 1/sqrt(2*num_decoder_layers).
```
Comment (RissyRan) [inline src/maxtext/configs/base.yml:399]: Shall we mention `1/sqrt(2*num_decoder_layers)` to align with Megatron?

File: src/maxtext/utils/grad_mask_utils.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+the configured threshold. Healthy tokens pass through unchanged. Used at
+decoder-layer boundaries to bound per-layer cotangent magnitudes
+(see deepseek model usage)."""
+
+import jax
+import jax.numpy as jnp
+
+
+@jax.custom_vjp
+def _grad_mask(x: jax.Array, threshold: jax.Array) -> jax.Array:
```
Comment (RissyRan) [inline src/maxtext/utils/grad_mask_utils.py:27]: Nit, maybe add `nondiff_argnums`?

```
import jax
import jax.numpy as jnp
from functools import partial

# Tell JAX that argument 1 (threshold) is not differentiable
@partial(jax.custom_vjp, nondiff_argnums=(1,))
def _grad_mask(x: jax.Array, threshold: float) -> jax.Array:
  return x

def _grad_mask_fwd(x: jax.Array, threshold: float):
  # No need to pass threshold as a residual since it's static/non-differentiable
  return x, ()

def _grad_mask_bwd(threshold: float, res, g: jax.Array):
  rms = jnp.sqrt(jnp.mean(jnp.square(g.astype(jnp.float32)), axis=-1, keepdims=True))
  mask = rms <= threshold
  # Only return the gradient for `x` (tuple of length 1)
  return (jnp.where(mask, g, jnp.zeros_like(g)),)

_grad_mask.defvjp(_grad_mask_fwd, _grad_mask_bwd)

def maybe_grad_mask(x: jax.Array, cfg) -> jax.Array:
  if cfg.grad_mask_threshold > 0.0:
    # No need to use jnp.asarray here
    return _grad_mask(x, float(cfg.grad_mask_threshold))
  return x
```

**Review by RissyRan — COMMENTED** [review]
Thanks for the tests! Just a few minor comments.
